docs/api.md

```diff
@@ -99,3 +99,38 @@ You can find sample requests/responses [here](./api-samples/v1/import.md)
 | ----- | -------- |
 | 200   | |
 | 500 | something went wrong (see logs of the simulator server) |
+
+## Watch the simulator's resources
+
+Monitor simulator resources and get updates automatically.
+
+### HTTP Request
+
+`GET /api/v1/watchresources`
+
+#### Parameter
+You must specify the `lastResourceVersion` of each resource, which can be retrieved using the `list` function of each resource.
+If you won't specify it, the simulator failed to watch the resource.
+
+|parameter|requirement|description|
+| ----- | --- | -------- |
+|podsLastResourceVersion|MUST|integer|
+|nodesLastResourceVersion|MUST|integer|
+|pvsLastResourceVersion|MUST|integer|
+|pvcsLastResourceVersion|MUST|integer|
+|scsLastResourceVersion|MUST|integer|
+|pcsLastResourceVersion|MUST|integer|
```

Comment:
hmm, I don't think so.

Our `/watchresources` API use `NewRetryWatcher ` in internal logic. 
The `NewRetryWatcher` will return error if we pass the "0" to a `lastResourceVersion`. If these values are `optional`, Isn't this likely to be a cause of error?
https://github.com/kubernetes/client-go/blob/3b969f96803febcac81f3a4272a9914270fbb3a3/tools/watch/retrywatcher.go#L66

Also, we shouldn't modify these value.
> Resource versions must be treated as opaque by clients and passed unmodified back to the server.
https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions


Therefore, I think it is generally expected that this `watch API` will be called after the `list API` is called. And at least, I'd liket to require it in this simulator.


BTW, `string` might be better than `integer`.
> You must not assume resource versions are numeric or collatable.

Reply:
> Our /watchresources API use NewRetryWatcher in internal logic.
The NewRetryWatcher will return error if we pass the "0" to a lastResourceVersion. If these values are optional, Isn't this likely to be a cause of error?

That makes sense.

But, still the API looks unkind to clients. 
What I expect to see in this API is something like `kubectl get Pods -w`. (= After listing/getting the requested object, watch for changes.)

How about deleting these last resource versions params completely? 
In most use cases, the clients need to list all resources directly from kube-apiserver and then call `/watchresources` after that right? 
If so, we don't need to let clients send the last resource versions to server. Just listing all supported objects and start to watch for changes looks more kind to clients.

Reply:
> If so, we don't need to let clients send the last resource versions to server. Just listing all supported objects and start to watch for changes looks more kind to clients.

Good question.
Your proposed feature is already provided as this method.
https://pkg.go.dev/k8s.io/client-go/tools/cache#NewInformer

I use `NewRetryWatcher` because it allows me to specify the lastresourceversion at the start of the watch. 

In the future, we will be able to operate several simulators at once.
At that time, it is not desirable to keep WATCHING all simulators.
That is, `watch` needs to stop once, and it needs to be possible to resume `watch` from the middle of the previous one.
And then, In order to restart the watch from the middle, the lastresourceversion needs to be specifiable from the client.

Reply:
My new idea was to make the lastresourcecersion optional parameter, and if not specified, to run the `list` on the server side, return the result as an `Added` event to the client, and then start the watch. 

WDYT?

Reply:
See also this article.
https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes

Reply:
Ah~, okay. I think now I fully understand your idea.

> My new idea was to make the lastresourcecersion optional parameter, and if not specified, to run the list on the server side, return the result as an Added event to the client, and then start the watch.

After understood your idea, I guess your previous idea, making lastresourcecersion required and using `/watchresources` only to notice the update, makes some sense. 

But, hmmmm, it's a difficult question. I think there are pros/cons for each strategy.

### making lastresourcecersion optional 

This is your new idea. 

**pros**
web UI only need to communicate with `/watchresources` API to fetch the resources. Not need to call kube-apiserver's list API anymore.
That makes the web UI's implementation much simpler, since the client performs the same logic regardless of whether it is the first time access or not. Just accessing `/watchresources` API is always fine.

**cons**
When a user accesses the web UI for the first time, web UI needs to build the resource's view by processing many `Added` event for all resources **one by one**. 
So, If there are too many resources, the building may take some time.

### making lastresourcecersion required 

This is your previous idea.

**pros**

When a user accesses the web UI for the first time, web UI calls list API of kube-apiserver.
By processing another logic for the first time access, the time that takes for first view will be decreased especially when the number of resources is large.

**cons**

Web UI needs to have two logics to show resources; one is for first-time access, and another is for `/watchresources `.

Reply:
So, to summarize, your new idea vs your previous idea looks like performance vs maintainability.